File: ios/ProseCreatorTTS/ProseCreatorTTS/TTSEngine.swift
// TTSEngine.swift
// Manages the sherpa-onnx Kokoro TTS model lifecycle and audio generation.
// Handles model loading from downloaded files, synthesis, and WAV encoding.

import Foundation
import AVFoundation

/// Kokoro voice preset definitions matching the mlx-audio/Kokoro naming convention.
struct KokoroVoice: Codable, Identifiable {
    let id: String        // e.g. "af_sky"
    let name: String      // e.g. "Sky"
    let gender: String    // "female" or "male"
    let speakerId: Int32  // 0-based index into Kokoro's speaker embedding
    let accent: String    // e.g. "American", "British"
}

/// All 25 Kokoro v0.19 voice presets.
/// The sherpa-onnx Kokoro v0.19 model packs voices into voices.bin.
/// Speaker IDs 0..10 correspond to the first 11 voices; the rest are accessed
/// by name through the voices.bin lookup when using newer sherpa-onnx builds.
/// For maximum compatibility, we map all 25 to their known speaker IDs.
let kokoroVoices: [KokoroVoice] = [
    // American Female
    KokoroVoice(id: "af_alloy", name: "Alloy", gender: "female", speakerId: 0, accent: "American"),
    KokoroVoice(id: "af_aoede", name: "Aoede", gender: "female", speakerId: 1, accent: "American"),
    KokoroVoice(id: "af_bella", name: "Bella", gender: "female", speakerId: 2, accent: "American"),
    KokoroVoice(id: "af_heart", name: "Heart", gender: "female", speakerId: 3, accent: "American"),
    KokoroVoice(id: "af_jessica", name: "Jessica", gender: "female", speakerId: 4, accent: "American"),
    KokoroVoice(id: "af_kore", name: "Kore", gender: "female", speakerId: 5, accent: "American"),
    KokoroVoice(id: "af_nicole", name: "Nicole", gender: "female", speakerId: 6, accent: "American"),
    KokoroVoice(id: "af_nova", name: "Nova", gender: "female", speakerId: 7, accent: "American"),
    KokoroVoice(id: "af_river", name: "River", gender: "female", speakerId: 8, accent: "American"),
    KokoroVoice(id: "af_sarah", name: "Sarah", gender: "female", speakerId: 9, accent: "American"),
    KokoroVoice(id: "af_sky", name: "Sky", gender: "female", speakerId: 10, accent: "American"),
    // American Male
    KokoroVoice(id: "am_adam", name: "Adam", gender: "male", speakerId: 11, accent: "American"),
    KokoroVoice(id: "am_echo", name: "Echo", gender: "male", speakerId: 12, accent: "American"),
    KokoroVoice(id: "am_eric", name: "Eric", gender: "male", speakerId: 13, accent: "American"),
    KokoroVoice(id: "am_liam", name: "Liam", gender: "male", speakerId: 14, accent: "American"),
    KokoroVoice(id: "am_michael", name: "Michael", gender: "male", speakerId: 15, accent: "American"),
    KokoroVoice(id: "am_onyx", name: "Onyx", gender: "male", speakerId: 16, accent: "American"),
    // British Female
    KokoroVoice(id: "bf_alice", name: "Alice", gender: "female", speakerId: 17, accent: "British"),
    KokoroVoice(id: "bf_emma", name: "Emma", gender: "female", speakerId: 18, accent: "British"),
    KokoroVoice(id: "bf_isabella", name: "Isabella", gender: "female", speakerId: 19, accent: "British"),
    KokoroVoice(id: "bf_lily", name: "Lily", gender: "female", speakerId: 20, accent: "British"),
    // British Male
    KokoroVoice(id: "bm_daniel", name: "Daniel", gender: "male", speakerId: 21, accent: "British"),
    KokoroVoice(id: "bm_fable", name: "Fable", gender: "male", speakerId: 22, accent: "British"),
    KokoroVoice(id: "bm_george", name: "George", gender: "male", speakerId: 23, accent: "British"),
    KokoroVoice(id: "bm_lewis", name: "Lewis", gender: "male", speakerId: 24, accent: "British"),
]

/// Look up a Kokoro voice by its ID string.
func kokoroVoiceByID(_ voiceID: String) -> KokoroVoice? {
    return kokoroVoices.first { $0.id == voiceID }
}

// MARK: - WAV Encoder

/// Encode a raw float PCM buffer into a WAV file in memory.
/// - Parameters:
///   - samples: Interleaved float samples in [-1.0, 1.0].
///   - sampleRate: Sample rate in Hz (e.g. 24000).
///   - channels: Number of audio channels (1 for mono).
/// - Returns: Data containing a complete WAV file.
func encodeWAV(samples: [Float], sampleRate: Int32, channels: Int16 = 1) -> Data {
    let bitsPerSample: Int16 = 16
    let bytesPerSample = Int(bitsPerSample) / 8
    let dataSize = samples.count * bytesPerSample * Int(channels)
    let fileSize = 36 + dataSize

    var data = Data()
    data.reserveCapacity(44 + dataSize)

    // RIFF header
    data.append(contentsOf: "RIFF".utf8)
    data.append(contentsOf: withUnsafeBytes(of: UInt32(fileSize).littleEndian) { Array($0) })
    data.append(contentsOf: "WAVE".utf8)

    // fmt sub-chunk
    data.append(contentsOf: "fmt ".utf8)
    data.append(contentsOf: withUnsafeBytes(of: UInt32(16).littleEndian) { Array($0) })    // SubChunk1Size
    data.append(contentsOf: withUnsafeBytes(of: UInt16(1).littleEndian) { Array($0) })     // PCM format
    data.append(contentsOf: withUnsafeBytes(of: UInt16(channels).littleEndian) { Array($0) })
    data.append(contentsOf: withUnsafeBytes(of: UInt32(sampleRate).littleEndian) { Array($0) })
    let byteRate = UInt32(sampleRate) * UInt32(channels) * UInt32(bytesPerSample)
    data.append(contentsOf: withUnsafeBytes(of: byteRate.littleEndian) { Array($0) })
    let blockAlign = UInt16(channels) * UInt16(bytesPerSample)
    data.append(contentsOf: withUnsafeBytes(of: blockAlign.littleEndian) { Array($0) })
    data.append(contentsOf: withUnsafeBytes(of: UInt16(bitsPerSample).littleEndian) { Array($0) })

    // data sub-chunk
    data.append(contentsOf: "data".utf8)
    data.append(contentsOf: withUnsafeBytes(of: UInt32(dataSize).littleEndian) { Array($0) })

    // Convert float samples to 16-bit PCM
    for sample in samples {
        let clamped = max(-1.0, min(1.0, sample))
        let intSample = Int16(clamped * Float(Int16.max))
        data.append(contentsOf: withUnsafeBytes(of: intSample.littleEndian) { Array($0) })
    }

    return data
}

// MARK: - TTS Engine

/// Errors that can occur during TTS engine operations.
enum TTSEngineError: LocalizedError {
    case modelNotLoaded
    case modelFileMissing(String)
    case initializationFailed
    case generationFailed
    case invalidVoice(String)

    var errorDescription: String? {
        switch self {
        case .modelNotLoaded:
            return "TTS model is not loaded. Download the model first."
        case .modelFileMissing(let file):
            return "Required model file missing: \(file)"
        case .initializationFailed:
            return "Failed to initialize the TTS engine. Check model files."
        case .generationFailed:
            return "Audio generation failed."
        case .invalidVoice(let voice):
            return "Unknown voice ID: \(voice)"
        }
    }
}

/// The TTS engine wraps sherpa-onnx's Kokoro model for on-device speech synthesis.
/// Thread-safe: all generation goes through a serial dispatch queue.
@MainActor
final class TTSEngine: ObservableObject {
    /// Whether the model is loaded and ready to generate.
    @Published var isLoaded: Bool = false
    /// Human-readable status message.
    @Published var statusMessage: String = "Not loaded"
    /// Loading progress (0.0 to 1.0) during model initialization.
    @Published var loadProgress: Double = 0.0

    private var ttsWrapper: SherpaOnnxOfflineTtsWrapper?
    private let generationQueue = DispatchQueue(label: "com.prosecreator.tts.generation", qos: .userInitiated)

    /// The sample rate of the loaded model (0 if not loaded).
    var sampleRate: Int32 {
        return ttsWrapper?.sampleRate ?? 0
    }

    /// Number of speakers in the loaded model.
    var numSpeakers: Int32 {
        return ttsWrapper?.numSpeakers ?? 0
    }

    /// Directory where Kokoro model files are stored after download.
    static var modelDirectory: URL {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        return docs.appendingPathComponent("kokoro-en-v0_19", isDirectory: true)
    }

    /// Check if all required model files exist on disk.
    static var modelFilesExist: Bool {
        let dir = modelDirectory
        let fm = FileManager.default
        let requiredFiles = ["model.onnx", "tokens.txt", "voices.bin"]
        for file in requiredFiles {
            if !fm.fileExists(atPath: dir.appendingPathComponent(file).path) {
                return false
            }
        }
        // espeak-ng-data directory must exist
        var isDir: ObjCBool = false
        let espeakPath = dir.appendingPathComponent("espeak-ng-data").path
        if !fm.fileExists(atPath: espeakPath, isDirectory: &isDir) || !isDir.boolValue {
            return false
        }
        return true
    }

    /// Load the Kokoro model from disk. Must be called after model download completes.
    func loadModel() async throws {
        statusMessage = "Loading model..."
        loadProgress = 0.1

        let dir = TTSEngine.modelDirectory
        let modelPath = dir.appendingPathComponent("model.onnx").path
        let voicesPath = dir.appendingPathComponent("voices.bin").path
        let tokensPath = dir.appendingPathComponent("tokens.txt").path
        let espeakDataDir = dir.appendingPathComponent("espeak-ng-data").path

        // Verify all files exist
        let fm = FileManager.default
        guard fm.fileExists(atPath: modelPath) else {
            throw TTSEngineError.modelFileMissing("model.onnx")
        }
        guard fm.fileExists(atPath: voicesPath) else {
            throw TTSEngineError.modelFileMissing("voices.bin")
        }
        guard fm.fileExists(atPath: tokensPath) else {
            throw TTSEngineError.modelFileMissing("tokens.txt")
        }
        guard fm.fileExists(atPath: espeakDataDir) else {
            throw TTSEngineError.modelFileMissing("espeak-ng-data/")
        }

        loadProgress = 0.3
        statusMessage = "Initializing sherpa-onnx..."

        // Build the config chain. We must keep C strings alive during init,
        // so we use a synchronous approach with withCString nesting.
        let wrapper: SherpaOnnxOfflineTtsWrapper? = await withCheckedContinuation { continuation in
            // Run on background thread since model loading can take several seconds
            generationQueue.async {
                let result = modelPath.withCString { modelCStr in
                    voicesPath.withCString { voicesCStr in
                        tokensPath.withCString { tokensCStr in
                            espeakDataDir.withCString { dataDirCStr in
                                "".withCString { emptyCStr -> SherpaOnnxOfflineTtsWrapper? in
                                    var kokoroConfig = SherpaOnnxOfflineTtsKokoroModelConfig()
                                    kokoroConfig.model = modelCStr
                                    kokoroConfig.voices = voicesCStr
                                    kokoroConfig.tokens = tokensCStr
                                    kokoroConfig.data_dir = dataDirCStr
                                    kokoroConfig.length_scale = 1.0
                                    kokoroConfig.dict_dir = emptyCStr
                                    kokoroConfig.lexicon = emptyCStr
                                    kokoroConfig.lang = emptyCStr

                                    var modelConfig = SherpaOnnxOfflineTtsModelConfig()
                                    modelConfig.kokoro = kokoroConfig

                                    var ttsConfig = SherpaOnnxOfflineTtsConfig()
                                    ttsConfig.model = modelConfig
                                    ttsConfig.rule_fsts = emptyCStr
                                    ttsConfig.max_num_sentences = 2
                                    ttsConfig.rule_fars = emptyCStr
                                    ttsConfig.silence_scale = 1.0

                                    return SherpaOnnxOfflineTtsWrapper(config: &ttsConfig)
                                }
                            }
                        }
                    }
                }
                continuation.resume(returning: result)
            }
        }

        guard let wrapper = wrapper else {
            loadProgress = 0.0
            statusMessage = "Failed to load model"
            throw TTSEngineError.initializationFailed
        }

        self.ttsWrapper = wrapper
        self.isLoaded = true
        self.loadProgress = 1.0
        self.statusMessage = "Model loaded (\(wrapper.sampleRate) Hz, \(wrapper.numSpeakers) speakers)"
    }

    /// Unload the model to free memory.
    func unloadModel() {
        ttsWrapper = nil
        isLoaded = false
        statusMessage = "Not loaded"
        loadProgress = 0.0
    }

    /// Generate speech audio from text.
    /// - Parameters:
    ///   - text: The text to synthesize.
    ///   - voice: Kokoro voice ID (e.g. "af_sky"). Defaults to "af_sky".
    ///   - speed: Speed multiplier. 1.0 = normal.
    /// - Returns: WAV file data.
    func generateSpeech(text: String, voice: String = "af_sky", speed: Float = 1.0) async throws -> Data {
        guard let wrapper = ttsWrapper, isLoaded else {
            throw TTSEngineError.modelNotLoaded
        }

        // Resolve voice to speaker ID
        let speakerId: Int32
        if let voicePreset = kokoroVoiceByID(voice) {
            speakerId = voicePreset.speakerId
        } else if let intId = Int32(voice) {
            // Allow numeric speaker IDs directly
            speakerId = intId
        } else {
            // Default to speaker 0 (af_alloy) if voice not found
            speakerId = 0
        }

        // Clamp speed to reasonable range
        let clampedSpeed = max(0.5, min(2.0, speed))

        let result: Data? = await withCheckedContinuation { continuation in
            generationQueue.async {
                guard let audio = wrapper.generate(text: text, sid: speakerId, speed: clampedSpeed) else {
                    continuation.resume(returning: nil)
                    return
                }

                let samples = audio.samplesArray()
                guard !samples.isEmpty else {
                    continuation.resume(returning: nil)
                    return
                }

                let wavData = encodeWAV(samples: samples, sampleRate: audio.sampleRate)
                continuation.resume(returning: wavData)
            }
        }

        guard let wavData = result else {
            throw TTSEngineError.generationFailed
        }

        return wavData
    }

    /// Estimate generation time for a given text length (rough heuristic).
    func estimateGenerationTime(textLength: Int) -> TimeInterval {
        // Kokoro on iPhone typically processes ~50 chars/second on A15+
        return TimeInterval(textLength) / 50.0
    }
}

File: android/app/src/main/java/ai/adverant/prosecreator/tts/TTSEngine.kt
package ai.adverant.prosecreator.tts

import android.util.Log
import com.k2fsa.sherpa.onnx.OfflineTts
import com.k2fsa.sherpa.onnx.OfflineTtsConfig
import com.k2fsa.sherpa.onnx.OfflineTtsKokoroModelConfig
import com.k2fsa.sherpa.onnx.OfflineTtsModelConfig
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * TTSEngine — Wraps sherpa-onnx OfflineTts for Kokoro model inference.
 *
 * Uses OfflineTtsKokoroModelConfig (NOT VitsModelConfig) for Kokoro-82M.
 * Generates PCM audio encoded as WAV (16-bit, 24000 Hz mono).
 *
 * Voice selection uses speaker IDs mapped from voices.bin. The voices
 * are embedded in alphabetical order within the single voices.bin file.
 *
 * Thread-safe: synthesize() is synchronized.
 */
class TTSEngine private constructor(
    private val tts: OfflineTts,
    private val modelDir: File,
) {

    companion object {
        private const val TAG = "TTSEngine"
        const val SAMPLE_RATE = 24000
        private const val BITS_PER_SAMPLE = 16
        private const val NUM_CHANNELS = 1

        /**
         * All Kokoro voice presets — sorted alphabetically.
         * Speaker IDs in voices.bin follow this exact order (0-indexed).
         */
        val KOKORO_VOICES: Map<String, String> = linkedMapOf(
            "af_alloy" to "Alloy (American Female)",
            "af_aoede" to "Aoede (American Female)",
            "af_bella" to "Bella (American Female)",
            "af_heart" to "Heart (American Female)",
            "af_jessica" to "Jessica (American Female)",
            "af_kore" to "Kore (American Female)",
            "af_nicole" to "Nicole (American Female)",
            "af_nova" to "Nova (American Female)",
            "af_river" to "River (American Female)",
            "af_sarah" to "Sarah (American Female)",
            "af_sky" to "Sky (American Female)",
            "am_adam" to "Adam (American Male)",
            "am_echo" to "Echo (American Male)",
            "am_eric" to "Eric (American Male)",
            "am_fenrir" to "Fenrir (American Male)",
            "am_liam" to "Liam (American Male)",
            "am_michael" to "Michael (American Male)",
            "am_onyx" to "Onyx (American Male)",
            "am_puck" to "Puck (American Male)",
            "bf_alice" to "Alice (British Female)",
            "bf_emma" to "Emma (British Female)",
            "bf_isabella" to "Isabella (British Female)",
            "bf_lily" to "Lily (British Female)",
            "bm_daniel" to "Daniel (British Male)",
            "bm_fable" to "Fable (British Male)",
            "bm_george" to "George (British Male)",
            "bm_lewis" to "Lewis (British Male)",
        )

        /** Sorted voice IDs — index = speaker ID in voices.bin. */
        private val SORTED_VOICE_IDS: List<String> = KOKORO_VOICES.keys.sorted()

        /** Default voice for narration */
        const val DEFAULT_VOICE = "af_sky"

        /**
         * Create a TTSEngine from model files in the given directory.
         * Returns null if initialization fails.
         */
        fun create(modelDir: File): TTSEngine? {
            return try {
                val modelPath = File(modelDir, "model.onnx")
                val tokensPath = File(modelDir, "tokens.txt")
                val voicesPath = File(modelDir, "voices.bin")
                val dataDir = File(modelDir, "espeak-ng-data")

                // Verbose file existence and size logging
                Log.i(TAG, "=== TTSEngine Initialization ===")
                Log.i(TAG, "Model dir: ${modelDir.absolutePath}")
                Log.i(TAG, "  model.onnx: exists=${modelPath.exists()}, size=${modelPath.length()} bytes (${modelPath.length() / 1_000_000}MB)")
                Log.i(TAG, "  tokens.txt: exists=${tokensPath.exists()}, size=${tokensPath.length()} bytes")
                Log.i(TAG, "  voices.bin: exists=${voicesPath.exists()}, size=${voicesPath.length()} bytes (${voicesPath.length() / 1_000}KB)")
                Log.i(TAG, "  espeak-ng-data: exists=${dataDir.exists()}, isDir=${dataDir.isDirectory}")

                if (dataDir.exists() && dataDir.isDirectory) {
                    val espeakFiles = dataDir.listFiles()?.map { "${it.name} (${it.length()})" } ?: emptyList()
                    Log.i(TAG, "  espeak-ng-data contents (${espeakFiles.size} files): $espeakFiles")
                } else {
                    Log.e(TAG, "  espeak-ng-data MISSING or not a directory — phonemizer will fail!")
                }

                // Also check for individual voice .bin files (wrong format)
                val binFiles = modelDir.listFiles()?.filter { it.name.endsWith(".bin") && it.name != "voices.bin" }
                if (!binFiles.isNullOrEmpty()) {
                    Log.w(TAG, "  Found ${binFiles.size} individual .bin files (wrong format for this config): ${binFiles.take(5).map { it.name }}")
                }

                if (!modelPath.exists()) {
                    Log.e(TAG, "FATAL: Model file not found: ${modelPath.absolutePath}")
                    return null
                }
                if (!tokensPath.exists()) {
                    Log.e(TAG, "FATAL: Tokens file not found: ${tokensPath.absolutePath}")
                    return null
                }
                if (!voicesPath.exists()) {
                    Log.w(TAG, "WARNING: voices.bin not found — speaker selection will not work")
                }
                if (!dataDir.exists()) {
                    Log.w(TAG, "WARNING: espeak-ng-data not found — phonemizer may fail, audio may be garbled")
                }

                val kokoroConfig = OfflineTtsKokoroModelConfig(
                    model = modelPath.absolutePath,
                    voices = if (voicesPath.exists()) voicesPath.absolutePath else "",
                    tokens = tokensPath.absolutePath,
                    dataDir = if (dataDir.exists()) dataDir.absolutePath else "",
                    lexicon = "",
                    lang = "en",
                    dictDir = "",
                    lengthScale = 1.0f,
                )

                Log.i(TAG, "KokoroConfig: model=${kokoroConfig.model}")
                Log.i(TAG, "KokoroConfig: voices=${kokoroConfig.voices}")
                Log.i(TAG, "KokoroConfig: tokens=${kokoroConfig.tokens}")
                Log.i(TAG, "KokoroConfig: dataDir=${kokoroConfig.dataDir}")
                Log.i(TAG, "KokoroConfig: lang=${kokoroConfig.lang}")

                val modelConfig = OfflineTtsModelConfig(
                    kokoro = kokoroConfig,
                    numThreads = 2,
                    debug = true, // Enable debug for diagnostics
                    provider = "cpu",
                )

                val config = OfflineTtsConfig(
                    model = modelConfig,
                    maxNumSentences = 1,
                    ruleFsts = "",
                    ruleFars = "",
                )

                Log.i(TAG, "Creating OfflineTts instance...")
                val tts = OfflineTts(config = config)
                Log.i(TAG, "TTSEngine initialized: sampleRate=${tts.sampleRate()}, numSpeakers=${tts.numSpeakers()}")

                if (tts.numSpeakers() == 0) {
                    Log.e(TAG, "WARNING: numSpeakers=0 — voices.bin may not have loaded correctly")
                }

                val engine = TTSEngine(tts, modelDir)

                // Run a quick test synthesis to verify the engine works
                Log.i(TAG, "Running test synthesis...")
                val testResult = engine.synthesizeWithDiagnostics("Hello", DEFAULT_VOICE, 1.0f)
                if (testResult.first != null) {
                    Log.i(TAG, "Test synthesis OK: ${testResult.first!!.size} bytes")
                    val diag = testResult.second
                    Log.i(TAG, "Test audio stats: ${diag["audio_stats"]}")
                } else {
                    Log.e(TAG, "Test synthesis FAILED — engine may not produce valid audio")
                }

                Log.i(TAG, "=== TTSEngine Ready ===")
                engine
            } catch (e: Exception) {
                Log.e(TAG, "Failed to initialize TTSEngine", e)
                null
            }
        }
    }

    /**
     * Synthesize text to WAV audio bytes.
     */
    @Synchronized
    fun synthesize(text: String, voice: String = DEFAULT_VOICE, speed: Float = 1.0f): ByteArray? {
        return synthesizeWithDiagnostics(text, voice, speed).first
    }

    /**
     * Synthesize text and return both WAV bytes and diagnostic info.
     */
    @Synchronized
    fun synthesizeWithDiagnostics(
        text: String,
        voice: String = DEFAULT_VOICE,
        speed: Float = 1.0f,
    ): Pair<ByteArray?, Map<String, Any>> {
        val diagnostics = mutableMapOf<String, Any>()
        diagnostics["voice_requested"] = voice
        diagnostics["speed"] = speed
        diagnostics["text_length"] = text.length
        diagnostics["text_preview"] = text.take(100)

        if (text.isBlank()) {
            diagnostics["error"] = "blank_text"
            return Pair(null, diagnostics)
        }

        return try {
            val speakerId = resolveVoiceSpeakerId(voice)
            diagnostics["speaker_id"] = speakerId
            diagnostics["voice_resolved"] = SORTED_VOICE_IDS.getOrElse(speakerId) { "unknown" }

            Log.d(TAG, "Synthesizing: voice=$voice sid=$speakerId speed=$speed text=${text.take(80)}")

            val startTime = System.nanoTime()
            val audio = tts.generateWithCallback(
                text = text,
                sid = speakerId,
                speed = speed,
                callback = { _ -> 0 }
            )
            val elapsedMs = (System.nanoTime() - startTime) / 1_000_000

            diagnostics["synthesis_time_ms"] = elapsedMs
            diagnostics["sample_rate"] = audio.sampleRate
            diagnostics["num_samples"] = audio.samples.size

            if (audio.samples.isEmpty()) {
                Log.w(TAG, "Empty audio generated for text: ${text.take(80)}")
                diagnostics["error"] = "empty_audio"
                return Pair(null, diagnostics)
            }

            // Audio quality diagnostics
            val stats = analyzeAudioSamples(audio.samples)
            diagnostics["audio_stats"] = stats
            Log.d(TAG, "Audio stats: $stats")

            // Warn if audio looks garbled
            val zeroRatio = stats["zero_ratio"] as? Float ?: 0f
            val maxAbs = stats["max_abs"] as? Float ?: 0f
            val rms = stats["rms"] as? Float ?: 0f

            if (zeroRatio > 0.95f) {
                Log.w(TAG, "GARBLED WARNING: >95% samples are zero — likely model failure")
                diagnostics["quality_warning"] = "mostly_silence"
            }
            if (maxAbs < 0.001f) {
                Log.w(TAG, "GARBLED WARNING: max amplitude < 0.001 — nearly silent")
                diagnostics["quality_warning"] = "nearly_silent"
            }
            if (rms > 0.9f) {
                Log.w(TAG, "GARBLED WARNING: RMS > 0.9 — audio is clipping/noise")
                diagnostics["quality_warning"] = "clipping_noise"
            }

            val durationSec = audio.samples.size.toFloat() / audio.sampleRate
            diagnostics["audio_duration_s"] = durationSec

            val wavBytes = encodeWav(audio.samples, audio.sampleRate)
            diagnostics["wav_size_bytes"] = wavBytes.size

            Log.i(TAG, "Synthesized ${wavBytes.size} bytes (${String.format("%.1f", durationSec)}s) in ${elapsedMs}ms for voice=$voice sid=$speakerId")

            Pair(wavBytes, diagnostics)
        } catch (e: Exception) {
            Log.e(TAG, "Synthesis failed for voice=$voice speed=$speed text=${text.take(80)}", e)
            diagnostics["error"] = e.message ?: "unknown_exception"
            diagnostics["error_type"] = e.javaClass.simpleName
            Pair(null, diagnostics)
        }
    }

    /**
     * Get comprehensive engine diagnostics for the /diagnostics endpoint.
     */
    fun getDiagnostics(): Map<String, Any> {
        val diag = mutableMapOf<String, Any>()
        diag["engine"] = "sherpa-onnx"
        diag["model_dir"] = modelDir.absolutePath
        diag["sample_rate"] = tts.sampleRate()
        diag["num_speakers"] = tts.numSpeakers()
        diag["available_voices"] = KOKORO_VOICES.size
        diag["sorted_voice_ids"] = SORTED_VOICE_IDS

        // Model file details
        val modelFiles = mutableMapOf<String, Any>()
        val filesToCheck = listOf("model.onnx", "tokens.txt", "voices.bin")
        for (fname in filesToCheck) {
            val f = File(modelDir, fname)
            modelFiles[fname] = mapOf(
                "exists" to f.exists(),
                "size_bytes" to f.length(),
                "size_mb" to String.format("%.1f", f.length() / 1_000_000.0),
            )
        }

        // espeak-ng-data directory
        val espeakDir = File(modelDir, "espeak-ng-data")
        if (espeakDir.exists() && espeakDir.isDirectory) {
            val files = espeakDir.listFiles() ?: emptyArray()
            val dirContents = mutableListOf<Map<String, Any>>()
            for (f in files) {
                if (f.isFile) {
                    dirContents.add(mapOf("name" to f.name, "size" to f.length()))
                } else if (f.isDirectory) {
                    val subCount = f.listFiles()?.size ?: 0
                    dirContents.add(mapOf("name" to "${f.name}/", "files" to subCount))
                }
            }
            modelFiles["espeak-ng-data"] = mapOf(
                "exists" to true,
                "file_count" to files.size,
                "contents" to dirContents,
            )
        } else {
            modelFiles["espeak-ng-data"] = mapOf("exists" to false)
        }

        // Check for stray individual .bin voice files (wrong format)
        val strayBins = modelDir.listFiles()
            ?.filter { it.name.endsWith(".bin") && it.name != "voices.bin" }
            ?.map { it.name } ?: emptyList()
        if (strayBins.isNotEmpty()) {
            modelFiles["stray_voice_bins"] = strayBins
        }

        // All files in model directory
        val allFiles = modelDir.listFiles()?.map { "${it.name} (${it.length()})" } ?: emptyList()
        diag["all_model_dir_files"] = allFiles

        diag["model_files"] = modelFiles
        return diag
    }

    fun getSampleRate(): Int = tts.sampleRate()

    fun getNumSpeakers(): Int = tts.numSpeakers()

    fun getAvailableVoices(): Map<String, String> = KOKORO_VOICES

    fun isVoiceAvailable(voiceId: String): Boolean = KOKORO_VOICES.containsKey(voiceId)

    fun release() {
        try {
            tts.release()
            Log.i(TAG, "TTSEngine released")
        } catch (e: Exception) {
            Log.w(TAG, "Error releasing TTSEngine", e)
        }
    }

    fun resolveVoiceSpeakerId(voice: String): Int {
        val idx = SORTED_VOICE_IDS.indexOf(voice)
        return if (idx >= 0) idx else {
            Log.w(TAG, "Unknown voice '$voice', falling back to $DEFAULT_VOICE")
            SORTED_VOICE_IDS.indexOf(DEFAULT_VOICE).coerceAtLeast(0)
        }
    }

    /**
     * Analyze audio samples for quality diagnostics.
     */
    private fun analyzeAudioSamples(samples: FloatArray): Map<String, Any> {
        if (samples.isEmpty()) return mapOf("empty" to true)

        var min = Float.MAX_VALUE
        var max = Float.MIN_VALUE
        var sum = 0.0
        var sumSq = 0.0
        var zeroCount = 0
        var nanCount = 0
        var infCount = 0
        var zeroCrossings = 0
        var prevSign = samples[0] >= 0

        for (i in samples.indices) {
            val s = samples[i]
            if (s.isNaN()) { nanCount++; continue }
            if (s.isInfinite()) { infCount++; continue }
            if (s < min) min = s
            if (s > max) max = s
            sum += s
            sumSq += s * s
            if (abs(s) < 0.0001f) zeroCount++

            val sign = s >= 0
            if (i > 0 && sign != prevSign) zeroCrossings++
            prevSign = sign
        }

        val n = samples.size
        val mean = sum / n
        val rms = sqrt(sumSq / n).toFloat()

        return mapOf(
            "num_samples" to n,
            "min" to min,
            "max" to max,
            "max_abs" to maxOf(abs(min), abs(max)),
            "mean" to String.format("%.6f", mean),
            "rms" to rms,
            "zero_ratio" to (zeroCount.toFloat() / n),
            "nan_count" to nanCount,
            "inf_count" to infCount,
            "zero_crossings" to zeroCrossings,
            "zero_crossing_rate" to (zeroCrossings.toFloat() / n),
            "first_10_samples" to samples.take(10).map { String.format("%.4f", it) },
            "last_10_samples" to samples.takeLast(10).map { String.format("%.4f", it) },
        )
    }

    private fun encodeWav(samples: FloatArray, sampleRate: Int): ByteArray {
        val pcmBytes = samples.size * 2
        val wavSize = 44 + pcmBytes

        val buffer = ByteBuffer.allocate(wavSize).order(ByteOrder.LITTLE_ENDIAN)

        buffer.put("RIFF".toByteArray())
        buffer.putInt(wavSize - 8)
        buffer.put("WAVE".toByteArray())

        buffer.put("fmt ".toByteArray())
        buffer.putInt(16)
        buffer.putShort(1)
        buffer.putShort(NUM_CHANNELS.toShort())
        buffer.putInt(sampleRate)
        buffer.putInt(sampleRate * NUM_CHANNELS * BITS_PER_SAMPLE / 8)
        buffer.putShort((NUM_CHANNELS * BITS_PER_SAMPLE / 8).toShort())
        buffer.putShort(BITS_PER_SAMPLE.toShort())

        buffer.put("data".toByteArray())
        buffer.putInt(pcmBytes)

        for (sample in samples) {
            val clamped = sample.coerceIn(-1.0f, 1.0f)
            val intVal = (clamped * 32767.0f).toInt().toShort()
            buffer.putShort(intVal)
        }

        return buffer.array()
    }
}
